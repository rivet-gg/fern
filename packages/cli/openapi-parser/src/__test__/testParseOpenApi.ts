import { AbsoluteFilePath, join, RelativeFilePath } from "@fern-api/fs-utils";
import { CONSOLE_LOGGER } from "@fern-api/logger";
import { serialization } from "@fern-api/openapi-ir-sdk";
import { createMockTaskContext } from "@fern-api/task-context";
import { parse, SpecImportSettings } from "../parse";

const FIXTURES_PATH = join(AbsoluteFilePath.of(__dirname), RelativeFilePath.of("fixtures"));

// eslint-disable-next-line jest/no-export
export function testParseOpenAPI(
    fixtureName: string,
    openApiFilename: string,
    asyncApiFilename?: string,
    settings?: SpecImportSettings
): void {
    // eslint-disable-next-line jest/valid-title
    describe(fixtureName, () => {
        it("parse open api", async () => {
            const absolutePathToOpenAPI = join(
                FIXTURES_PATH,
                RelativeFilePath.of(fixtureName),
                RelativeFilePath.of(openApiFilename)
            );
            const absolutePathToAsyncAPI =
                asyncApiFilename != null
                    ? join(FIXTURES_PATH, RelativeFilePath.of(fixtureName), RelativeFilePath.of(asyncApiFilename))
                    : undefined;
            const specs = [];
            if (absolutePathToOpenAPI != null) {
                specs.push({
                    absoluteFilepath: absolutePathToOpenAPI,
                    absoluteFilepathToOverrides: undefined,
                    settings
                });
            }
            if (absolutePathToAsyncAPI != null) {
                specs.push({
                    absoluteFilepath: absolutePathToAsyncAPI,
                    absoluteFilepathToOverrides: undefined,
                    settings
                });
            }

            const openApiIr = await parse({
                specs,
                taskContext: createMockTaskContext({ logger: CONSOLE_LOGGER })
            });
            const openApiIrJson = await serialization.OpenApiIntermediateRepresentation.jsonOrThrow(openApiIr, {
                skipValidation: true
            });
            expect(openApiIrJson).toMatchSnapshot();
        });
    });
}
