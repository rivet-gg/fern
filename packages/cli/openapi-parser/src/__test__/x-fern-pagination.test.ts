import { testParseOpenAPI } from "./testParseOpenApi";

describe("x-fern-pagination", () => {
    testParseOpenAPI("x-fern-pagination", "openapi.yml");
});
