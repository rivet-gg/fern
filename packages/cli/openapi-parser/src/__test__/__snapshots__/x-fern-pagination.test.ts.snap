// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`x-fern-pagination x-fern-pagination parse open api 1`] = `
{
  "apiVersion": null,
  "basePath": null,
  "channel": [],
  "description": null,
  "endpoints": [
    {
      "audiences": [],
      "authed": false,
      "availability": null,
      "description": "",
      "errors": {},
      "examples": [
        {
          "codeSamples": [],
          "description": null,
          "headers": [],
          "name": null,
          "pathParameters": [],
          "queryParameters": [],
          "request": null,
          "response": {
            "properties": {
              "users": {
                "type": "array",
                "value": [
                  {
                    "properties": {
                      "age": {
                        "type": "primitive",
                        "value": {
                          "type": "int",
                          "value": 1,
                        },
                      },
                      "email": {
                        "type": "primitive",
                        "value": {
                          "type": "string",
                          "value": "email",
                        },
                      },
                      "id": {
                        "type": "primitive",
                        "value": {
                          "type": "string",
                          "value": "id",
                        },
                      },
                      "name": {
                        "type": "primitive",
                        "value": {
                          "type": "string",
                          "value": "name",
                        },
                      },
                    },
                    "type": "object",
                  },
                ],
              },
            },
            "type": "object",
          },
          "type": "full",
        },
      ],
      "generatedRequestName": "ListRequest",
      "headers": [],
      "idempotent": null,
      "internal": null,
      "method": "GET",
      "operationId": "list",
      "pagination": {
        "offset": "$request.pageNumber",
        "results": "$response.users",
        "step": null,
        "type": "offset",
      },
      "path": "/list",
      "pathParameters": [],
      "queryParameters": [
        {
          "availability": null,
          "description": null,
          "name": "pageNumber",
          "parameterNameOverride": null,
          "schema": {
            "availability": null,
            "description": null,
            "generatedName": "ListRequestPageNumber",
            "groupName": null,
            "nameOverride": null,
            "type": "nullable",
            "value": {
              "availability": null,
              "description": null,
              "generatedName": "ListRequestPageNumber",
              "groupName": null,
              "nameOverride": null,
              "schema": {
                "default": null,
                "exclusiveMaximum": null,
                "exclusiveMinimum": null,
                "maximum": null,
                "minimum": null,
                "multipleOf": null,
                "type": "int",
              },
              "type": "primitive",
            },
          },
        },
      ],
      "request": null,
      "requestNameOverride": null,
      "response": {
        "description": "Success",
        "responseProperty": null,
        "schema": {
          "availability": null,
          "description": null,
          "generatedName": "ListResponse",
          "groupName": null,
          "nameOverride": null,
          "schema": "UsersResponse",
          "type": "reference",
        },
        "type": "json",
      },
      "sdkName": null,
      "server": [],
      "summary": "List",
      "tags": [],
    },
    {
      "audiences": [],
      "authed": false,
      "availability": null,
      "description": "",
      "errors": {},
      "examples": [
        {
          "codeSamples": [],
          "description": null,
          "headers": [],
          "name": null,
          "pathParameters": [],
          "queryParameters": [],
          "request": null,
          "response": {
            "properties": {
              "users": {
                "type": "array",
                "value": [
                  {
                    "properties": {
                      "age": {
                        "type": "primitive",
                        "value": {
                          "type": "int",
                          "value": 1,
                        },
                      },
                      "email": {
                        "type": "primitive",
                        "value": {
                          "type": "string",
                          "value": "email",
                        },
                      },
                      "id": {
                        "type": "primitive",
                        "value": {
                          "type": "string",
                          "value": "id",
                        },
                      },
                      "name": {
                        "type": "primitive",
                        "value": {
                          "type": "string",
                          "value": "name",
                        },
                      },
                    },
                    "type": "object",
                  },
                ],
              },
            },
            "type": "object",
          },
          "type": "full",
        },
      ],
      "generatedRequestName": "ListRequest",
      "headers": [],
      "idempotent": null,
      "internal": null,
      "method": "GET",
      "operationId": "list",
      "pagination": {
        "offset": "$request.pageNumber",
        "results": "$response.users",
        "step": "$request.limit",
        "type": "offset",
      },
      "path": "/list-with-step",
      "pathParameters": [],
      "queryParameters": [
        {
          "availability": null,
          "description": null,
          "name": "pageNumber",
          "parameterNameOverride": null,
          "schema": {
            "availability": null,
            "description": null,
            "generatedName": "ListRequestPageNumber",
            "groupName": null,
            "nameOverride": null,
            "type": "nullable",
            "value": {
              "availability": null,
              "description": null,
              "generatedName": "ListRequestPageNumber",
              "groupName": null,
              "nameOverride": null,
              "schema": {
                "default": null,
                "exclusiveMaximum": null,
                "exclusiveMinimum": null,
                "maximum": null,
                "minimum": null,
                "multipleOf": null,
                "type": "int",
              },
              "type": "primitive",
            },
          },
        },
        {
          "availability": null,
          "description": null,
          "name": "limit",
          "parameterNameOverride": null,
          "schema": {
            "availability": null,
            "description": null,
            "generatedName": "ListRequestLimit",
            "groupName": null,
            "nameOverride": null,
            "type": "nullable",
            "value": {
              "availability": null,
              "description": null,
              "generatedName": "ListRequestLimit",
              "groupName": null,
              "nameOverride": null,
              "schema": {
                "default": null,
                "exclusiveMaximum": null,
                "exclusiveMinimum": null,
                "maximum": null,
                "minimum": null,
                "multipleOf": null,
                "type": "int",
              },
              "type": "primitive",
            },
          },
        },
      ],
      "request": null,
      "requestNameOverride": null,
      "response": {
        "description": "Success",
        "responseProperty": null,
        "schema": {
          "availability": null,
          "description": null,
          "generatedName": "ListResponse",
          "groupName": null,
          "nameOverride": null,
          "schema": "UsersResponse",
          "type": "reference",
        },
        "type": "json",
      },
      "sdkName": null,
      "server": [],
      "summary": "List",
      "tags": [],
    },
  ],
  "globalHeaders": [],
  "groups": {},
  "hasEndpointsMarkedInternal": false,
  "idempotencyHeaders": [],
  "nonRequestReferencedSchemas": [
    "UsersResponse",
    "User",
  ],
  "schemas": {
    "User": {
      "additionalProperties": false,
      "allOf": [],
      "allOfPropertyConflicts": [],
      "availability": null,
      "description": null,
      "generatedName": "User",
      "groupName": null,
      "nameOverride": null,
      "properties": [
        {
          "audiences": [],
          "availability": null,
          "conflict": {},
          "generatedName": "userId",
          "key": "id",
          "nameOverride": null,
          "schema": {
            "availability": null,
            "description": null,
            "generatedName": "userId",
            "groupName": null,
            "nameOverride": null,
            "type": "optional",
            "value": {
              "availability": null,
              "description": null,
              "generatedName": "UserId",
              "groupName": null,
              "nameOverride": null,
              "schema": {
                "default": null,
                "format": "uuid",
                "maxLength": null,
                "minLength": null,
                "pattern": null,
                "type": "string",
              },
              "type": "primitive",
            },
          },
        },
        {
          "audiences": [],
          "availability": null,
          "conflict": {},
          "generatedName": "userName",
          "key": "name",
          "nameOverride": null,
          "schema": {
            "availability": null,
            "description": null,
            "generatedName": "userName",
            "groupName": null,
            "nameOverride": null,
            "type": "optional",
            "value": {
              "availability": null,
              "description": null,
              "generatedName": "UserName",
              "groupName": null,
              "nameOverride": null,
              "schema": {
                "default": null,
                "format": null,
                "maxLength": null,
                "minLength": null,
                "pattern": null,
                "type": "string",
              },
              "type": "primitive",
            },
          },
        },
        {
          "audiences": [],
          "availability": null,
          "conflict": {},
          "generatedName": "userEmail",
          "key": "email",
          "nameOverride": null,
          "schema": {
            "availability": null,
            "description": null,
            "generatedName": "userEmail",
            "groupName": null,
            "nameOverride": null,
            "type": "optional",
            "value": {
              "availability": null,
              "description": null,
              "generatedName": "UserEmail",
              "groupName": null,
              "nameOverride": null,
              "schema": {
                "default": null,
                "format": null,
                "maxLength": null,
                "minLength": null,
                "pattern": null,
                "type": "string",
              },
              "type": "primitive",
            },
          },
        },
        {
          "audiences": [],
          "availability": null,
          "conflict": {},
          "generatedName": "userAge",
          "key": "age",
          "nameOverride": null,
          "schema": {
            "availability": null,
            "description": null,
            "generatedName": "userAge",
            "groupName": null,
            "nameOverride": null,
            "type": "optional",
            "value": {
              "availability": null,
              "description": null,
              "generatedName": "UserAge",
              "groupName": null,
              "nameOverride": null,
              "schema": {
                "default": null,
                "exclusiveMaximum": null,
                "exclusiveMinimum": null,
                "maximum": null,
                "minimum": null,
                "multipleOf": null,
                "type": "int",
              },
              "type": "primitive",
            },
          },
        },
      ],
      "type": "object",
    },
    "UsersResponse": {
      "additionalProperties": false,
      "allOf": [],
      "allOfPropertyConflicts": [],
      "availability": null,
      "description": null,
      "generatedName": "UsersResponse",
      "groupName": null,
      "nameOverride": null,
      "properties": [
        {
          "audiences": [],
          "availability": null,
          "conflict": {},
          "generatedName": "usersResponseUsers",
          "key": "users",
          "nameOverride": null,
          "schema": {
            "availability": null,
            "description": null,
            "generatedName": "usersResponseUsers",
            "groupName": null,
            "nameOverride": null,
            "type": "optional",
            "value": {
              "availability": null,
              "description": null,
              "generatedName": "UsersResponseUsers",
              "groupName": null,
              "nameOverride": null,
              "type": "array",
              "value": {
                "availability": null,
                "description": null,
                "generatedName": "UsersResponseUsersItem",
                "groupName": null,
                "nameOverride": null,
                "schema": "User",
                "type": "reference",
              },
            },
          },
        },
      ],
      "type": "object",
    },
  },
  "securitySchemes": {},
  "servers": [],
  "tags": {
    "orderedTagIds": null,
    "tagsById": {},
  },
  "title": "Test extension \`x-fern-pagination\`",
  "variables": {},
  "webhooks": [],
}
`;
